students/TacSPx/lesson04/mailroompart2.py

```diff
@@ -118,46 +110,37 @@ def letter(donor_name, donor_amount):
 XYZ Nonprofit Agency Director
 '''
     return new_letter
+
+
 ###############################################
 #   #2 Create a Report
 ###############################################
+def report_sorting(donor_data):  # added new report sorting removed from report function to clean-up code
+    ''' report_sorting: organizes and calculates data for report function'''
+    sorter = []
+    for keys, values in donor_data.items():
+        sorter.append([sum(values), keys, len(values), sum(values) / len(values)])
+    return sorted(sorter, reverse=True)
+
+
 def report():
-    ''' Report: creates a tabular summary report of donor transactions'''
+    ''' report: creates a tabular summary report of donor transactions'''
     while True:
-        donor_sort = {}
-        donor_sum = {}
-        donor_len = {}
-        donor_avg = {}
-        donor_merge = {}
-        # title
         title = "*******************  XYZ Nonprofit Charity  ********************\n" \
                 "________________________________________________________________"
         heading_name = ("| Donor Name:", "| Totals $:", "| #-of-Gifts:", "| Avg Gift $: ", "|")
         header = f"{heading_name[0].upper():<20s} {heading_name[1].upper():<13s}" \
                  f" {heading_name[2].upper():<6s} {heading_name[3].upper():<14s}|"
         print(title)
         print(header)
-        # define format row
-        row = "| {Name:<18s} | ${Total:<10.2f} | {NumbGifts:^11d} | ${AverageGifts:<10.2f} |".format
-        # revisions made below to add sorting by totals, step #1 set-up to sort, new list
-
-        # calcs in the dictionary
-        donor_sum = dict([(key, sum(values)) for key, values in donor_info.items()])
-        donor_len = dict([(key, len(values)) for key, values in donor_info.items()])
-        donor_avg = dict([(key, sum(values) / len(values)) for key, values in donor_info.items()])
-        # merge
-        for key in set(list(donor_sum.keys()) + list(donor_len.keys()) + list(donor_avg.keys())):
-            donor_merge.setdefault(key, []).append(donor_sum[key])
-            donor_merge.setdefault(key, []).append(donor_len[key])
-            donor_merge.setdefault(key, []).append(donor_avg[key])
-        # sort dictionary by totals
-        donor_sort = sorted(donor_merge.items(), key=itemgetter(1), reverse=True)
-        # report printing
-        for i in donor_sort:
-            print(row(Name=i[0], Total=i[1][0], NumbGifts=i[1][1], AverageGifts=i[1][2]))
+        # revised and split off sorting process into new report_sorting function
+        call_sorting = report_sorting(donor_info)
+        for i in range(0, len(call_sorting)):
+            print(f"| {call_sorting[i][1]:<18s} | ${call_sorting[i][0]:<10.2f} | {call_sorting[i][2]:^11d}"
+                  f" | ${call_sorting[i][3]:<10.2f} |")
```

Comment:
the sorting is much simpler than that, see below:
```
def sort_key(item):
    return sum(item[1])


sorted_donors = sorted(donor_info.items(), key=sort_key, reverse=True)
```
items() creates a list of tuples (key + value), sort_key will get an item from a list and you will return the sum of the second item which is the donations.

then simply loop over (you don't need to use range here) the sorted list:
```
for donor, donations in sorted_donor:
    print(...)
```